dkg/frost/frost.go

```diff
@@ -0,0 +1,756 @@
+	signer types.DKGSigner,
+	storage dkg.Storage,
+) dkg.KeyGenProtocol {
+
+	msgs := make(map[DKGRound]map[uint32]*dkg.SignedMessage)
+	msgs[Preparation] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Round1] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Round2] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Blame] = make(map[uint32]*dkg.SignedMessage)
+
+	return &FROST{
+		identifier: requestID,
+		network:    network,
+		signer:     signer,
+		storage:    storage,
+		operatorID: operatorID,
+
+		msgs:           msgs,
+		operatorShares: make(map[uint32]*bls.SecretKey),
+	}
+}
+
+func (fr *FROST) Start(init *dkg.Init) error {
+	otherOperators := make([]uint32, 0)
+	for _, operatorID := range init.OperatorIDs {
+		if fr.operatorID == operatorID {
+			continue
+		}
+		otherOperators = append(otherOperators, uint32(operatorID))
+	}
+
+	operators := []uint32{uint32(fr.operatorID)}
+	operators = append(operators, otherOperators...)
+	fr.operators = operators
+
+	pctx := make([]byte, 16)
+	_, err := rand.Read(pctx)
+	if err != nil {
+		return err
+	}
+
+	party, err := frost.NewDkgParticipant(uint32(fr.operatorID), uint32(len(operators)), string(pctx), thisCurve, otherOperators...)
+	if err != nil {
+		return err
+	}
+
+	fr.party = party
+	fr.threshold = uint32(init.Threshold)
+
+	k, err := ecies.GenerateKey()
+	if err != nil {
+		return err
+	}
+	fr.sessionSK = k
+
+	fr.currentRound = Preparation
+
+	protocolMessage := &ProtocolMsg{
+		Round: Preparation,
+		PreparationMessage: &PreparationMessage{
+			SessionPk: k.PublicKey.Bytes(true),
+		},
+	}
+
+	protocolMessageBytes, err := protocolMessage.Encode()
+	if err != nil {
+		return err
+	}
+
+	bcastPrepMessage := &dkg.SignedMessage{
+		Message: &dkg.Message{
+			MsgType:    dkg.ProtocolMsgType,
+			Identifier: fr.identifier,
+			Data:       protocolMessageBytes,
+		},
+		Signer:    types.OperatorID(fr.operatorID),
+		Signature: nil,
+	}
+
+	exist, operator, err := fr.storage.GetDKGOperator(fr.operatorID)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		return errors.Errorf("operator with id %d not found", fr.operatorID)
+	}
+
+	sig, err := fr.signer.SignDKGOutput(bcastPrepMessage, operator.ETHAddress)
+	if err != nil {
+		return err
+	}
+
+	bcastPrepMessage.Signature = sig
+	fr.msgs[Preparation][uint32(fr.operatorID)] = bcastPrepMessage
+
+	return fr.network.BroadcastDKGMessage(bcastPrepMessage)
+}
+
+func (fr *FROST) ProcessMsg(msg *dkg.SignedMessage) (bool, *dkg.KeyGenOutput, error) {
+	if err := msg.Validate(); err != nil {
+		return false, nil, errors.Wrap(err, "failed to validate message signature")
+	}
+
+	protocolMessage := &ProtocolMsg{}
+	if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+		return false, nil, errors.Wrap(err, "could not decode protocol msg")
+	}
+
+	if fr.msgs[protocolMessage.Round] == nil {
+		fr.msgs[protocolMessage.Round] = make(map[uint32]*dkg.SignedMessage)
+	}
+
+	originalMessage, ok := fr.msgs[protocolMessage.Round][uint32(msg.Signer)]
```

Comment:
why does this make it inconsistent? what if we simply get the same message twice?